Tests/LibCompress/TestLzma.cpp

```diff
@@ -0,0 +1,214 @@
+    auto decompressor = MUST(Compress::LzmaDecompressor::create(move(stream), options));
+    auto buffer_or_error = decompressor->read_until_eof(PAGE_SIZE);
+    EXPECT(buffer_or_error.is_error());
+}
+
+TEST_CASE(specification_bad_eos_incorrect_size_lzma_decompress)
+{
+    Compress::LzmaDecompressorOptions options {
+        .literal_context_bits = 3,
+        .literal_position_bits = 0,
+        .position_bits = 2,
+        .dictionary_size = 65536,
+        .uncompressed_size = 328,
+    };
+    Array<u8, 109> const compressed {
+        0x00, 0x26, 0x16, 0x85, 0xBC, 0x45, 0xF0, 0xDF, 0xFF, 0xD2, 0xE8, 0x41, 0xF5, 0xCE, 0xE5, 0x90,
+        0xE1, 0xC8, 0x20, 0xEA, 0xC6, 0x37, 0xBE, 0x2B, 0xD1, 0xF4, 0xC3, 0x34, 0x6F, 0x2F, 0x83, 0xC2,
+        0xA6, 0x7C, 0x6F, 0x3D, 0x88, 0xA0, 0x58, 0x22, 0x1F, 0x3A, 0xBA, 0x7B, 0xC6, 0xDD, 0x66, 0xFE,
+        0xF8, 0x92, 0xE4, 0xCB, 0x1C, 0xC4, 0x19, 0x0A, 0x0C, 0x8B, 0x2E, 0x39, 0xB8, 0xB8, 0x03, 0xCD,
+        0x5A, 0x9E, 0x10, 0x3A, 0x4F, 0x65, 0xFA, 0x41, 0xCB, 0xF2, 0x79, 0x65, 0xD7, 0xF1, 0x9F, 0xAB,
+        0x70, 0x1D, 0x6F, 0xF7, 0xB6, 0x79, 0xCC, 0x8A, 0x7D, 0xCE, 0xDB, 0xF8, 0xF6, 0x9E, 0xC9, 0x12,
+        0x9F, 0xAA, 0xBF, 0x8A, 0x08, 0xF5, 0x99, 0x8D, 0x7F, 0xFA, 0x18, 0x0A, 0x52
+    };
+
+    auto stream = MUST(try_make<FixedMemoryStream>(compressed));
+    auto decompressor = MUST(Compress::LzmaDecompressor::create(move(stream), options));
+    auto buffer_or_error = decompressor->read_until_eof(PAGE_SIZE);
+    EXPECT(buffer_or_error.is_error());
+}
+
+TEST_CASE(specification_bad_incorrect_size_lzma_decompress)
+{
+    Compress::LzmaDecompressorOptions options {
+        .literal_context_bits = 3,
+        .literal_position_bits = 0,
+        .position_bits = 2,
+        .dictionary_size = 8388608,
+        .uncompressed_size = 290,
+    };
+    Array<u8, 104> const compressed {
+        0x00, 0x26, 0x16, 0x85, 0xBC, 0x45, 0xF0, 0xDF, 0xFF, 0xD2, 0xE8, 0x41, 0xF5, 0xCE, 0xE5, 0x90,
+        0xE1, 0xC8, 0x20, 0xEA, 0xC6, 0x37, 0xBE, 0x2B, 0xD1, 0xF4, 0xC3, 0x34, 0x6F, 0x2F, 0x83, 0xC2,
+        0xA6, 0x7C, 0x6F, 0x3D, 0x88, 0xA0, 0x58, 0x22, 0x1F, 0x3A, 0xBA, 0x7B, 0xC6, 0xDD, 0x66, 0xFE,
+        0xF8, 0x92, 0xE4, 0xCB, 0x1C, 0xC4, 0x19, 0x0A, 0x0C, 0x8B, 0x2E, 0x39, 0xB8, 0xB8, 0x03, 0xCD,
+        0x5A, 0x9E, 0x10, 0x3A, 0x4F, 0x65, 0xFA, 0x41, 0xCB, 0xF2, 0x79, 0x65, 0xD7, 0xF1, 0x9F, 0xAB,
+        0x70, 0x1D, 0x6F, 0xF7, 0xB6, 0x79, 0xCC, 0x8A, 0x7D, 0xCE, 0xDB, 0xF8, 0xF6, 0x9E, 0xC9, 0x12,
+        0x9F, 0xAA, 0xBF, 0x89, 0xFE, 0x05, 0x36, 0x80
+    };
+
+    auto stream = MUST(try_make<FixedMemoryStream>(compressed));
+    auto decompressor = MUST(Compress::LzmaDecompressor::create(move(stream), options));
+    // FIXME: This should detect that there is still remaining (non-'end of stream') data after the expected data and reject the last read.
+    //        As of now, this test accepts any result, except for crashes.
+    (void)decompressor->read_until_eof(PAGE_SIZE);
```

Comment:
> Are we going to fix this FIXME?

I plan to, yes. However, since my first attempt at this led to a bunch of erroneous failures and end-of-stream detection will have a significant effect on formats that embed LZMA, I decided to let it be for now and revisit it later, focusing on getting the main parts of the decoder (and the followup support for LZMA2 and XZ) finished up and merged first.

> Also, we might want to expose a is_corrupted helper that returns true if we reached the 'known' uncompressed size but m_range_decoder_code is non-0? The spec says that's a good indicator for an incorrectly decoded stream (and you already check that when we reach the explicit end indicator), so we can assert that returns true here.

The specification indeed mentions that this is a good way of checking for corruption, but it unfortunately doesn't seem to hold up if the archive contains an uncompressed size _and_ an end of stream marker (since it still has to read the end marker at that point). This is also what I suspect was causing my failures that shouldn't have been there.

In any case, the current mode of operation of this implementation is to stop and report an error even for things that would otherwise just be "corrupted", so I'd rather have it report that immediately if it even just notices that something is wrong.

Reply:
I see, let's leave that to a future PR then :+1: